src/components/Admin/Trainer/ControllerDataset.js

```diff
@@ -0,0 +1,202 @@
+ */
+import * as tf from '@tensorflow/tfjs';
+import firebase from 'firebase/app';
+import 'firebase/firestore';
+import 'firebase/storage';
+
+export class ControllerDataset {
+	constructor(status) {
+		this.setStatus = status;
+
+		firebase.initializeApp({
+			apiKey: 'AIzaSyBVuVNKx-rx2ON0RxbfGfbGPpiymbMrxj8',
+			authDomain: 'honesty-store-kiosk.firebaseapp.com',
+			projectId: 'honesty-store-kiosk'
+		});
+		this.store = firebase.storage();
+		this.db = firebase.firestore();
+		this.db.settings({timestampsInSnapshots: true});
+	}
+
+	async setItemTrainingCounts(itemObj) {
+		const items = await this.db.collection('item_data').get();
+		items.forEach(item => {
+			const count = item.data().count;
+			itemObj[item.id].mlCount = count ? count : 0;
+		});
+		return itemObj;
+	}
+
+	async addExample(img, activation, label) {
+		const item = await this.db
+			.collection('item_data')
+			.doc(label)
+			.get();
+
+		this.db
+			.collection('item_data')
+			.doc(label)
+			.set({
+				count: item.exists ? item.data().count + 1 : 1
+			});
+
+		this.db
+			.collection('training_data')
+			.add({
+				img,
+				activation: activation.dataSync().join(','),
+				label,
+				random: Math.random(),
+				trusted: true
+			})
+			.then(ref => {
+				activation.dispose();
+			})
+			.catch(err => {
+				activation.dispose();
+			});
+	}
+
+	async addExamples(examples) {
+		if (examples.length < 1) return;
+
+		const item = await this.db
+			.collection('item_data')
+			.doc(examples[0].label)
+			.get();
+
+		this.db
+			.collection('item_data')
+			.doc(examples[0].label)
+			.set({
+				count: item.exists
+					? item.data().count + examples.length
+					: examples.length
+			});
+
+		for (let i in examples) {
+			this.db
+				.collection('training_data')
+				.add({
+					img: examples[i].img,
+					activation: examples[i].activation.dataSync().join(','),
+					label: examples[i].label,
+					random: Math.random(),
+					trusted: true
+				})
+				.then(() => {
+					examples[i].activation.dispose();
+					this.setStatus(
+						`Uploading images (${parseInt(i, 10) + 1}/${
+							examples.length
+						})`
+					);
+					if (parseInt(i, 10) + 1 >= examples.length) {
+						this.setStatus('Done');
+					}
+				})
+				.catch(err => {
+					console.error(err);
+					examples[i].activation.dispose();
+				});
+		}
+	}
+
+	async getClasses() {
+		const items = await this.db.collection('item_data').get();
+		let idList = [];
+		items.forEach(doc => {
+			idList.push(doc.id);
+		});
+		return idList;
+	}
+
+	async getBatch(batchSize, randomness, since) {
+		return new Promise(async resolve => {
+			let batch = {};
+
+			while (Object.keys(batch).length < batchSize) {
+				const snapshot = await this.db
+					.collection('training_data')
+					.orderBy('timestamp')
+					.startAt(since)
+					.orderBy('random')
+					.startAt(Math.random())
+					.limit(batchSize * randomness)
+					.get();
+
+				await snapshot.forEach(doc => {
```

Comment:
I think the awaiting problem is coming from the forEach, see https://stackoverflow.com/questions/37576685/using-async-await-with-a-foreach-loop

We should be avoid to avoid the wrapper promise since `async/await` is just promises under the hood

```
  for (const doc of snapshot) {
    batch[doc.id] = await doc.data();
  }
```

Reply:
Snapshot isn't an array, it's QuerySnapshot returned from the DB so their implementation of forEach doesn't work quite the same as with an array. This was the best solution I found.